File: dart/lib/cell.dart
// Copyright (c) 2015, Kay-Uwe Kirstein. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

part of mazes;

/// Cell class a a single grid element
class Cell {
  int row, column;
  Cell north, south, east, west;
  Map<Cell, bool> _links;

  // constructor
  Cell(this.row, this.column) {
    _links = {};
  }

  // link/unlink to other cells
  link(other, [bidirect = true]) {
    _links[other] = true;
    if (bidirect) {
      other.link(this, false);
    }
  }

  unlink(other, [bidirect = true]) {
    _links.remove(other);
    if (bidirect) {
      other.unlink(this, false);
    }
  }

  links() {
    return _links.keys;
  }

  linked(other) {
    return _links.containsKey(other);
  }

  // get neighbors
  neighbors() {
    var list = [north, south, west, east];

    return list.where((e) => e != null);
  }

  // toString interface
  String toString() {
    return "Cell: row: ${row}, col: ${column}";
  }
}

File: ocaml/lib/cell.ml
(* vim: set ft=ocaml sw=2 ts=2 *)

(**
 * Module cell
 * A single cell
*)

type t = {
  row : int;
  col : int;
  mutable north : t option;
  mutable south : t option;
  mutable east : t option;
  mutable west : t option;
  mutable links : t Hashset.t
}

(**
 * Creates a cell for given position
*)
let create row col = {
  row; col;
  north = None; south = None; east = None; west = None;
  links = Hashset.create 64
}


(**
 * link two cell together
 *)
let link c1 c2 =
  Hashset.add c1.links c2;
  Hashset.add c2.links c1

(**
 * unlink two cells
 *)
let unlink c1 c2 =
  Hashset.remove c1.links c2;
  Hashset.remove c2.links c1

(**
 * predicate whether two cells are linked
 *)
let is_linked c1 c2 =
  Hashset.mem c1.links c2

(**
 * returns a list of linked cells
 *)
let links c =
  Hashset.fold (fun x acc -> x :: acc) c.links [] |>
  List.rev

(**
 * returns the linked neighbors
 *)
let neighbors c =
  [c.north; c.south; c.east; c.west] |>
  List.fold_left (fun acc x -> match x with Some c -> c :: acc | None -> acc) []


